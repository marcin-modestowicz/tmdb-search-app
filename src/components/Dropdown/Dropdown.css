.root {
  position: relative;
  display: inline-block;
  padding: 0 0.5em;
  cursor: pointer;
  border-radius: 0.25em;
  font-family: sans-serif;
  color: dimGray;
  border: 1px solid dimGray;
  font-size: 1em;
  line-height: 2em;
}

.root.active {
  background-color: silver;
  border-color: silver;
  color: white;
}

.options {
  position: absolute;
  left: 0;
  top: 2.25em;
  display: block;
  margin: 0;
  padding: 0;
  box-shadow: 0 0 0.25em silver;
  border-radius: 0.25em;
  max-height: 500px;
  width: 100%;
  z-index: 1000;
  list-style: none;
  min-width: 100px;
}

.option {
  padding: 0 0.5em;
  color: silver;
}

.option:first-child {
  border-radius: 0.25em 0.25em 0 0;
}

.option:last-child {
  border-radius: 0 0 0.25em 0.25em;
}

.option:hover {
  background-color: royalBlue;
  color: white;
}
