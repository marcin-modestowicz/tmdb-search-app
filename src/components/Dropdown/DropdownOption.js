// @flow

import React, { Component } from "react";
import styles from "./Dropdown.css";

type Props = {
  value: string,
  label: string,
  onClick: (value: string) => void
};

class DropdownOption extends Component<Props> {
  props: Props;

  handleClick = () => {
    this.props.onClick(this.props.value);
  };

  render() {
    return (
      <div role="button" className={styles.option} onClick={this.handleClick}>
        {this.props.label}
      </div>
    );
  }
}

export default DropdownOption;
