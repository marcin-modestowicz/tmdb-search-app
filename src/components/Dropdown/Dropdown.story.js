// @flow

import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import Dropdown from "./Dropdown";

storiesOf("Dropdown", module).add("simple", () => (
  <Dropdown
    value="one"
    options={[{ value: "one", label: "One" }, { value: "two", label: "Two" }]}
    onSelect={action("option selected")}
  />
));
