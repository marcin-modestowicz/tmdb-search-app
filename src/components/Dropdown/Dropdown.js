// @flow

import React, { Component } from "react";
import cn from "classnames";
import { reaction, observable, action, computed } from "mobx";
import { observer } from "mobx-react";
import type { FieldValue } from "../../types";
import DropdownOption from "./DropdownOption";
import styles from "./Dropdown.scss";

type Props = {
  value: any,
  options: FieldValue[],
  onSelect: (value: string) => void
};

@observer
class Dropdown extends Component<Props> {
  props: Props;

  static Option = DropdownOption;

  @observable isListFocused: boolean = false;

  @action
  handleListBlur = () => {
    this.isListFocused = false;
  };

  @action
  handleListFocus = () => {
    this.isListFocused = true;
  };

  @action
  handleClick = () => {
    this.isListFocused = true;
  };

  @action
  handleSelect = (value: string) => {
    this.isListFocused = false;

    if (this.props.onSelect) {
      this.props.onSelect(value);
    }
  };

  render() {
    const { value, options } = this.props;

    return (
      <div className={cn(styles.root, { [styles.active]: this.isListFocused })}>
        <div role="button" onClick={this.handleClick}>
          {value}
        </div>
        {options &&
          this.isListFocused && (
            <ul
              className={styles.options}
              onMouseEnter={this.handleListFocus}
              onMouseLeave={this.handleListBlur}
            >
              {options.map(({ value, label }) => (
                <Dropdown.Option
                  key={value}
                  value={value}
                  label={label}
                  onClick={this.handleSelect}
                />
              ))}
            </ul>
          )}
      </div>
    );
  }
}

export default Dropdown;
