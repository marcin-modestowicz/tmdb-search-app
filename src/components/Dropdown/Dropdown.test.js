// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Dropdown from "./Dropdown";

describe("Dropdown component", () => {
  let dropdown;

  beforeEach(() => {
    dropdown = shallow(
      <Dropdown
        value="one"
        options={[
          { value: "one", label: "One" },
          { value: "two", label: "Two" }
        ]}
        onSelect={jest.fn()}
      />
    );
  });

  test("should render without options", () => {
    expect(toJson(dropdown)).toMatchSnapshot();
  });

  test("should show options when clicked", () => {
    dropdown
      .find("div")
      .last()
      .simulate("click");

    expect(toJson(dropdown)).toMatchSnapshot();
  });

  test("should hide options when option is chosen", () => {
    dropdown
      .find("div")
      .last()
      .simulate("click");
    dropdown
      .find("ul")
      .children()
      .last()
      .simulate("click");

    expect(toJson(dropdown)).toMatchSnapshot();
  });

  test("should hide options when list is not active", () => {
    dropdown
      .find("div")
      .last()
      .simulate("click");
    dropdown.find("ul").simulate("mouseLeave");

    expect(toJson(dropdown)).toMatchSnapshot();
  });
});
