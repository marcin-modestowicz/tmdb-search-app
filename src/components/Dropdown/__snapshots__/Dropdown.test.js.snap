// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should hide options when list is not active 1`] = `
<div
  className="root"
>
  <div
    onClick={[Function]}
    role="button"
  >
    one
  </div>
</div>
`;

exports[`Dropdown component should hide options when option is chosen 1`] = `
<div
  className="root"
>
  <div
    onClick={[Function]}
    role="button"
  >
    one
  </div>
</div>
`;

exports[`Dropdown component should render without options 1`] = `
<div
  className="root"
>
  <div
    onClick={[Function]}
    role="button"
  >
    one
  </div>
</div>
`;

exports[`Dropdown component should show options when clicked 1`] = `
<div
  className="root active"
>
  <div
    onClick={[Function]}
    role="button"
  >
    one
  </div>
  <ul
    className="options"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <DropdownOption
      key="one"
      label="One"
      onClick={[Function]}
      value="one"
    />
    <DropdownOption
      key="two"
      label="Two"
      onClick={[Function]}
      value="two"
    />
  </ul>
</div>
`;
